handlers/cassandra/cassandra.go

```diff
@@ -37,168 +45,154 @@ var (
 	HistSetBufferWait        = metrics.AddHistogram("set_batch_buffer_timewait", false, nil)
 )
 
-var singleton *Handler
-
 // SetReadonlyMode switch Cassandra handler to readonly mode for graceful exit
-func SetReadonlyMode() {
-	singleton.readonlymode = true
+func (h *Handler) Shutdown() {
+	atomic.AddUint32(&h.isShutingDown, 1)
+	h.executors.Wait()
 }
 
-func bufferSizeCheckLoop() {
-	ticker := time.NewTicker(5 * time.Millisecond)
-	for {
-		select {
-		case <-ticker.C:
-			if len(singleton.setbuffer) >= viper.GetInt("CassandraBatchMinItemSize") {
-				go FlushBuffer()
-			}
-		}
-	}
+func (h *Handler) IsShutingDown() bool {
+	return atomic.LoadUint32(&h.isShutingDown) > 0
 }
 
-// FlushBuffer triggers a batched write operation in Cassandra target
-func FlushBuffer() {
-	chanLen := len(singleton.setbuffer)
-	metrics.SetIntGauge(MetricSetBufferSize, uint64(chanLen))
+// We read from the channel from N executor/goroutine in order to propagate the writes
+func cassandraSetReal(cassandraHandler *Handler) {
 
-	if chanLen > 0 {
-		metrics.IncCounter(MetricCmdSetBatch)
+	for !cassandraHandler.IsShutingDown() {
+		item := <-cassandraHandler.setbuffer
+		query := cassandraHandler.session.Query(cassandraHandler.setStmt, item.Key, item.Data, item.Exptime)
 
-		// fmt.Println(chanLen)
-		if chanLen >= viper.GetInt("CassandraBatchMaxItemSize") {
-			chanLen = viper.GetInt("CassandraBatchMaxItemSize")
-		}
-		b := singleton.session.NewBatch(gocql.UnloggedBatch)
-		for i := 1; i <= chanLen; i++ {
-			item := (<-singleton.setbuffer)
-			b.Query(
-				fmt.Sprintf(
-					"INSERT INTO %s.%s (keycol,valuecol) VALUES (?, ?) USING TTL ?",
-					viper.GetString("CassandraKeyspace"),
-					viper.GetString("CassandraBucket"),
-				),
-				item.Key,
-				item.Data,
-				item.Exptime,
-			)
+		if err := query.Exec(); err != nil {
+			log.Println("[ERROR] Cassandra SET returned an error. ", err)
 		}
 
-		// exec CQL batch
-		start := timer.Now()
-		err := singleton.session.ExecuteBatch(b)
-		if err != nil {
-			metrics.IncCounter(MetricCmdSetBatchErrors)
-			log.Println("[ERROR] Batched Cassandra SET returned an error. ", err)
-		} else {
-			metrics.IncCounter(MetricCmdSetBatchSuccess)
-			metrics.ObserveHist(HistSetBatch, timer.Since(start))
+		//batch := cassandraHandler.session.NewBatch(gocql.UnloggedBatch)
+		//for i := 1; i <= 1000; i++ {
+		//	item := <-cassandraHandler.setbuffer
+		//	batch.Query(queryStr, item.Key, item.Data, item.Exptime)
+		//}
+		//if err := cassandraHandler.session.ExecuteBatch(batch); err != nil {
+		//	log.Println("[ERROR] Cassandra BATCHED SET returned an error. ", err)
+		//}
+
+	}
+
+	// Exiting: Flush remaining requests
+	for {
+		select {
+		case item := <-cassandraHandler.setbuffer:
+			cassandraHandler.session.Query(cassandraHandler.setStmt, item.Key, item.Data, item.Exptime).Exec()
+
+		// We are empty
```

Comment:
You "we are empty" but the code break after a duration? Could you make the comment more explicit?